File: udp/send_local/python/udp_local_server.py
#!/usr/bin/env python3

import socket
from argparse import ArgumentParser

MAX_BYTES = 65535

def server(port):
    sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    sock.bind(('127.0.0.1', port))
    print('Listening at {}'.format(sock.getsockname()))
    while True:
        data, address = sock.recvfrom(MAX_BYTES)
        text = data.decode('ascii')
        print('The client at {} says {!r}'.format(address, text))
        text = 'Your data was {} bytes long'.format(len(data))
        data = text.encode('ascii')
        sock.sendto(data, address)

if __name__ == '__main__':
    parser = ArgumentParser(description='Send and receive UDP locally')
    parser.add_argument('--port', '-p', type=int, default=1060, help='UDP port (default 1060)')
    args = parser.parse_args()

    print("args: {}".format(args))

    server(args.port)

File: udp/send_local/cxx/udp_local_server.cpp
#include <cstring>
#include <string>
#include <sstream>
#include <iostream>
#include <gflags/gflags.h>
#include "wrapsock.hpp"

#define MAX_BYTES 65535

void server(uint16_t port) {
    Socket sock(AF_INET, SOCK_DGRAM);
    sock.Bind("127.0.0.1", port);
    std::cout << "Listening at " << sock.Getsockname().ToString() << std::endl;
    while (true) {
#if __cplusplus >= 201703L
        auto [text, address] = sock.RecvFrom(MAX_BYTES);
#else
        std::string text;
        SocketAddress address;
        std::tie(text, address) = sock.RecvFrom(MAX_BYTES);
#endif
        std::cout << "The client at " << address << " says " << text << std::endl;
        text = "Your data was " + std::to_string(text.size()) + " bytes long";
        sock.SendTo(text, address);
    }
}

std::string usage(const char* prog) {
    std::ostringstream os;
    os << "Send and receive UDP locally\n\n"
        << "Usage: " << prog << " [--help] [--port PORT]";
    return os.str();
}

DEFINE_int32(port, 1060, "TCP port number");

int main(int argc, char *argv[]) {
    gflags::SetUsageMessage(usage(argv[0]));
    gflags::ParseCommandLineFlags(&argc, &argv, true);

    server(FLAGS_port);

    return 0;
}

